onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -23,19 +53,71 @@ static std::string ConcatNames(
+// Collect the node-arg names and the correspnding consumers name for reporting
 static std::string GetName(const std::pair<const NodeArg*, std::vector<Node*>>& p) {
   return p.first->Name() + " feeding " + ConcatNames(p.second) + "; ";
 };
-// The collection fp16_allow_ops, specifies for a given propagate_cast_ops level, a vector of node op_types that
-// the code is allowed to propage Cast operations cross. The user may specify a custom list of optypes using level 0.
-// The opcodes are split into multiple levels. Cast propagation is done based on the level. Level 2 op code
-// list includes Level 1 list also.
+/*
+* The collection fp16_allow_ops, specifies for a given propagate_cast_ops level, a vector of node op_types that
+* the code is allowed to propage Cast operations across. The user may specify a custom list of optypes using level 0.
+* The opcodes are split into multiple levels. Cast propagation is done based on the level. Level 2 op code
+* list includes Level 1 list also.
+*/
 static std::vector<std::unordered_set<std::string>> fp16_allow_ops = {
     /* Level 0 */ {},
     /* Level 1 */ {"Transpose", "Relu", "Reshape", "Split", "Tanh"},
     /* Level 2 */ {"BiasGelu", "Dropout", "FastGelu", "Gather", "Gelu", "LayerNormalization", "Where"}};
+
+/*
+*  The following two maps pecify the opcode to input and opcode to output mappings to list the inputs/outputs to consider while propagating
+*  cast operations. All other inputs/outputs not listed in this table are not relevant for deciding whether an operation
+*  performed in float or float16. If an opcode is not listed in these tables, the code will look at all the inputs and outputs to validate 
+*  transformation.
+*/
+static std::unordered_map<std::string, std::vector<int>> opcode_to_input_map = {
+    {"Gather", {0}},
+    {"Reshape", {0}},
+    {"Dropout", {0}},
+    {"LayerNormalization", {0, 1, 2}},
+};
+
+static std::unordered_map<std::string, std::vector<int>> opcode_to_output_map = {
+    {"Gather", {0}},
+    {"Reshape", {0}},
+    {"Dropout", {0}},
+    {"LayerNormalization", {0}},
+};
+
 static std::unordered_set<std::string> inserted_node_names;   // Names of the nodes inserted
 static std::unordered_set<std::string> converted_node_names;  // Names of the nodes converted to FP16
+
+/*
+*  Check if the input is relevant to consider for cast propagation for the given node.
+*  Return true if the opcode is not found in the opcode_to_input map.
+*/
+static bool IsRelevantInput(const Node* node, const NodeArg* input) {
+  if (opcode_to_input_map.find(node->OpType()) != opcode_to_input_map.end()) {
+    const std::vector<int>& selected_inputs = opcode_to_input_map[node->OpType()];
+    int input_index = optimizer_utils::IndexOfNodeInput(*node, *input);
+    return std::find(selected_inputs.begin(), selected_inputs.end(), input_index) != selected_inputs.end();
+  }
+  return true;
+}
+
+/*
+*  Check if the outnput is relevant to consider for cast propagation for the given node.
```

Comment:
>outnput [](http://example.com/codeflow?start=16&length=7)

typo

Reply:
I am installing the visual studio spellchecker extension to avoid typos. :)